Implement Flist module

zaibon opened the pull request:
- [x] Implement Flister interface
  - [x] Mount
  - [x] Umount
- [x] Tests
   - [x] business logic
   - [x] RPC interface
- [x] Implement Zbus interface

zaibon committed:
add flist module

zaibon committed:
add .gitignore files

zaibon committed:
add makefile for go modules

zaibon committed:
remove flist module binary

zaibon committed:
add test with -race to the makefile

zaibon committed:
flist module: add some more test and fix isolation

- make the cache common for each mount but a different backend
- add some test for isolation

zaibon committed:
add -v to go test makefile target

zaibon committed:
flist module: be more restrictive on the path supported for umount

zaibon committed:
flist module: add rpc tests

zaibon committed:
Makefile: add generate target

zaibon committed:
flist module: add flags for msg broker and worker number

muhamadazmy commented on modules/flist/flist.go:
```diff
@@ -0,0 +1,235 @@
+		storage = defaultStorage
+	}
+
+	rc, err := downloadFlist(url)
```
Check how 0-core does it, because we also now cache the flists as well. so we don't download the same flist over and over.

zaibon replied:
Indeed, that is the plan, I still have a todo in the downloadFlist function for that.

muhamadazmy commented on modules/flist/flist.go:
```diff
@@ -0,0 +1,235 @@
+		"-cache", f.cache,
+		"-meta", flistPath,
+		"-storage-url", storage,
+		"-daemon",
```
`daemon` Is that a new flag? I can't find it in g8ufs implementation. Also do we really need to demonize g8ufs ? We can always just spawn the fs, then wait on both the pid file and that the mountpoint is ready.

In current 0-core the way we do it is that we wait for the "mount ready" on fuse stdout, then we know for sure that the mount is ready. I think this approach is better than waiting for 2 seconds

zaibon replied:
> daemon Is that a new flag? I can't find it in g8ufs implementation. Also do we really need to demonize g8ufs ? We can always just spawn the fs, then wait on both the pid file and that the mountpoint is ready.

I believe the have to demonize cause otherwise the g8ufs processes are stopped if the flist module is restarted.

> In current 0-core the way we do it is that we wait for the "mount ready" on fuse stdout, then we know for sure that the mount is ready. I think this approach is better than waiting for 2 seconds

Yep this is something I need to improve, there is a todo in the code for it.

zaibon committed:
flist module: Avoid to download flist that are already present locally

zaibon committed:
rename zbus to zbusc

zaibon committed:
Improve proper startup detection of 0-fs daemon

- module now wait for both pid and scan logs of the daemon to assest it
has started properly

zaibon committed:
split RPC test boilerplate into its own function

LeeSmet commented on modules/flist/flist.go:
```diff
@@ -0,0 +1,325 @@
+	backend := filepath.Join(f.root, "backend", name)
+	_ = os.RemoveAll(logPath)
+	_ = os.RemoveAll(backend)
+	_ = os.RemoveAll(path)
```
why are these errors ignored?

zaibon committed:
flist module: handle error during clean up of working dirs

zaibon committed:
flist module: add documentation how to run the tests

muhamadazmy commented on modules/flist/flist.go:
```diff
@@ -0,0 +1,328 @@
+// it uses a MultiWriter to write the flist in a temporary file and fill up
+// the md5 hash then it rename the file to the hash
+func (f *flistModule) saveFlist(r io.Reader) (string, error) {
+	tmp, err := ioutil.TempFile(f.flist, "*_flist_temp")
```
Please consider the size of the tmpfs on a zos system, we might not have enough space under /tmp.

zaibon replied:
I don't write it into `/tmp` here. The first argument of `TempFile` define the directory to use. Here it is part of the working directory of the module itself.

muhamadazmy commented on modules/flist/flist.go:
```diff
@@ -0,0 +1,328 @@
+		return "", err
+	}
+
+	if os.Rename(tmp.Name(), path); err != nil {
```
os.Rename will not work across devices. since /tmp is usually a different partition or filesystem other than the final flist destination witch is usually a permanent disk partition (or cache) Rename will  fail.

Copying a file though might cause a race condition if u start to mount 2 locations with same flist. This is why in core0 I used to lock the flist file on disk first before i start the download to make sure only one "mount" is downloading the file.

zaibon replied:
Like explained in the previous comment `tmp` file  is not located in `/tmp`. It is created in the same directory as its final location, so the Rename is an atomic operation. So we don't need locks and are safe from race conditions.

LeeSmet reviewed (approved):
LGTM
